core/include/detray/utils/root_finding.hpp

```diff
@@ -0,0 +1,302 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s).
+#include "detray/definitions/detail/algebra.hpp"
+#include "detray/definitions/detail/math.hpp"
+#include "detray/definitions/detail/qualifiers.hpp"
+#include "detray/definitions/units.hpp"
+#include "detray/utils/invalid_values.hpp"
+
+// System include(s).
+#include <algorithm>
+#include <array>
+#include <iostream>
+#include <limits>
+#include <stdexcept>
+#include <string>
+#include <tuple>
+
+namespace detray {
+
+/// @brief Try to find a bracket around a root
+///
+/// @param [in] a lower initial boundary
+/// @param [in] b upper initial boundary
+/// @param [in] f function for which to find the root
+/// @param [out] bracket bracket around the root
+/// @param [in] k scale factor with which to widen the bracket at every step
+///
+/// @see Numerical Recepies pp. 445
+///
+/// @return whether a bracket was found
+template <typename scalar_t, typename function_t>
+DETRAY_HOST_DEVICE inline bool expand_bracket(const scalar_t a,
+                                              const scalar_t b, function_t &f,
+                                              std::array<scalar_t, 2> &bracket,
+                                              const scalar_t k = 1.f) {
+
+    if (a == b) {
+        throw std::invalid_argument(
+            "Root bracketing: Not a valid start interval [" +
+            std::to_string(a) + ", " + std::to_string(b) + "]");
+    }
```

Comment:
Throwing exceptions is not supported in most device platforms; perhaps an assert would be better.

Reply:
Is there no way to throw a fatal error in device code?

Reply:
Not in a portable way... I'm a bit confused as to how this even compiled; is this being used in CUDA code anywhere?

Reply:
This code is only run on the host for now, but you are right, in the future it could be used on device. Mainly, I do not know how to use the random number generation for the helix construction in a thread safe way and/or in device code. I heard random number generators may not be statistically independent when seeded incorrectly across multiple threads and I did not want to deal with that right now.

Reply:
So, now you just have to wait for a few hours until this code is done doing its this in serial execution :D

Reply:
I will remove the exception later, but for now, I rely on it to check if the root finding had a fatal failure anywhere during automatic data generation